front/lib/api/assistant/generation.ts

```diff
@@ -255,6 +272,193 @@ export async function renderConversationForModel({
   });
 }
 
+export async function renderConversationForModelMultiActions({
+  conversation,
+  model,
+  prompt,
+  allowedTokenCount,
+}: {
+  conversation: ConversationType;
+  model: { providerId: string; modelId: string };
+  prompt: string;
+  allowedTokenCount: number;
+}): Promise<
+  Result<
+    {
+      modelConversation: ModelConversationTypeMultiActions;
+      tokensUsed: number;
+    },
+    Error
+  >
+> {
+  const messages: ModelMessageTypeMultiActions[] = [];
+
+  // Render all messages and all actions.
+  for (let i = conversation.content.length - 1; i >= 0; i--) {
+    const versions = conversation.content[i];
+    const m = versions[versions.length - 1];
+
+    if (isAgentMessageType(m)) {
+      if (m.content) {
+        messages.unshift({
+          role: "assistant" as const,
+          name: m.configuration.name,
+          content: m.content,
+        });
+      }
+      if (m.action) {
```

Comment:
We'll need `m.actions`  but will create a separate task for that. 

@lasryaric does your current work on inverting relationship include refactoring from `message.action` into `message.actions` ?